File: backend/internal/models/price_sync.go
package models

// price_sync.go — 文本及多媒体模型价格自动同步领域协议。
//
// 本文件只定义同步目标、外部价格、预览批次、应用结果和管理配置；
// 不访问数据库，不调用供应商，不参与图片、视频或TTS业务结算。
//
// 安全边界：
//   - 文本价格按精确模型名匹配；
//   - 媒体价格沿用现有五元组身份；
//   - 预览与应用严格分离；
//   - 定时任务只处理AutoSyncEnabled=true的目标；
//   - 手动应用必须明确选择“全部”或指定明细ID，禁止空选择误应用全部。

import (
	"encoding/json"
	"time"
)

const (
	PriceSyncSourceMainGateway     = "main_gateway"
	PriceSyncSourceDomesticGateway = "domestic_gateway"
	PriceSyncSourceMediaGateway    = "media_gateway"
	PriceSyncSourceTTSGateway      = "tts_gateway"
)

const (
	PriceSyncTriggerManual    = "manual"
	PriceSyncTriggerScheduler = "scheduler"
)

const (
	PriceSyncRunPreviewed = "previewed"
	PriceSyncRunApplied   = "applied"
	PriceSyncRunFailed    = "failed"
)

const (
	PriceSyncTargetText  = "text"
	PriceSyncTargetMedia = "media"
)

const (
	PriceSyncActionUpdate    = "update"
	PriceSyncActionUnchanged = "unchanged"
	PriceSyncActionSkipped   = "skipped"
	PriceSyncActionApplied   = "applied"
	PriceSyncActionStale     = "stale"
)

// TextPriceSyncTarget 是同步服务内部使用的文本价格目标。
type TextPriceSyncTarget struct {
	ID              string
	ModelName       string
	Provider        string
	CostPer1kInput  float64
	CostPer1kOutput float64
	DisplayName     string
	IsActive        bool

	AutoSyncEnabled bool
	SyncSource      string
	SyncModelName   string
	LastSyncedAt    *time.Time
	LastSyncStatus  string
	LastSyncMessage string
}

// MediaPriceSyncTarget 是同步服务内部使用的媒体价格目标。
type MediaPriceSyncTarget struct {
	ID              string
	MediaType       string
	Provider        string
	ModelName       string
	Variant         string
	MediaUnit       string
	UnitCostUSD     float64
	MinimumQuantity float64
	MinimumCostUSD  float64
	DisplayName     string
	IsActive        bool

	AutoSyncEnabled bool
	SyncSource      string
	SyncModelName   string
	LastSyncedAt    *time.Time
	LastSyncStatus  string
	LastSyncMessage string
}

// GatewayPrice 是价格来源返回的单个模型规范化记录。
type GatewayPrice struct {
	ModelName              string             `json:"model_name"`
	Provider               string             `json:"provider"`
	Description            string             `json:"description"`
	QuotaType              int                `json:"quota_type"`
	ModelRatio             float64            `json:"model_ratio"`
	CompletionRatio        float64            `json:"completion_ratio"`
	ModelPriceUSD          float64            `json:"model_price_usd"`
	EnableGroups           []string           `json:"enable_groups"`
	SupportedEndpointPaths []string           `json:"supported_endpoint_paths"`
	UnitPricesUSD          map[string]float64 `json:"unit_prices_usd"`
	RawPayload             json.RawMessage    `json:"raw_payload"`
}

// GatewayPriceCatalog 是一次价格接口读取结果。
type GatewayPriceCatalog struct {
	Source      string
	PricingURL  string
	GroupRatios map[string]float64
	Prices      []GatewayPrice
}

// PriceSyncRun 是一次同步预览或应用批次。
type PriceSyncRun struct {
	ID            string          `json:"id"`
	TriggerType   string          `json:"trigger_type"`
	Status        string          `json:"status"`
	SourceKind    string          `json:"source_kind"`
	SourceBaseURL string          `json:"source_base_url"`
	PreviewOnly   bool            `json:"preview_only"`
	Summary       json.RawMessage `json:"summary"`
	ErrorMessage  string          `json:"error_message"`
	CreatedBy     *string         `json:"created_by"`
	StartedAt     time.Time       `json:"started_at"`
	FinishedAt    *time.Time      `json:"finished_at"`
}

// PriceSyncItem 是单个本地价格与上游价格的比对结果。
type PriceSyncItem struct {
	ID         string `json:"id"`
	RunID      string `json:"run_id"`
	TargetKind string `json:"target_kind"`
	TargetID   string `json:"target_id"`

	Provider   string `json:"provider"`
	ModelName  string `json:"model_name"`
	SyncSource string `json:"sync_source"`

	MediaType string `json:"media_type"`
	Variant   string `json:"variant"`
	MediaUnit string `json:"media_unit"`

	OldInputUSD    float64 `json:"old_input_usd"`
	NewInputUSD    float64 `json:"new_input_usd"`
	OldOutputUSD   float64 `json:"old_output_usd"`
	NewOutputUSD   float64 `json:"new_output_usd"`
	OldUnitCostUSD float64 `json:"old_unit_cost_usd"`
	NewUnitCostUSD float64 `json:"new_unit_cost_usd"`

	Action        string          `json:"action"`
	Reason        string          `json:"reason"`
	SourcePayload json.RawMessage `json:"source_payload"`
	CreatedAt     time.Time       `json:"created_at"`
}

// PriceSyncSummary 汇总一次同步结果。
type PriceSyncSummary struct {
	TotalCount     int `json:"total_count"`
	UpdateCount    int `json:"update_count"`
	UnchangedCount int `json:"unchanged_count"`
	SkippedCount   int `json:"skipped_count"`
	AppliedCount   int `json:"applied_count"`
	StaleCount     int `json:"stale_count"`
}

// PriceSyncPreviewRequest 发起同步预览。
type PriceSyncPreviewRequest struct {
	TriggerType      string  `json:"trigger_type"`
	Group            string  `json:"group"`
	IncludeText      bool    `json:"include_text"`
	IncludeMedia     bool    `json:"include_media"`
	MaxChangePercent float64 `json:"max_change_percent"`
}

// PriceSyncPreviewResponse 返回同步批次及全部比对明细。
type PriceSyncPreviewResponse struct {
	Run     *PriceSyncRun    `json:"run"`
	Items   []PriceSyncItem  `json:"items"`
	Summary PriceSyncSummary `json:"summary"`
}

// PriceSyncApplyRequest 应用已预览的价格变化。
//
// ApplyAll=true表示应用本批全部update明细。
// ApplyAll=false时ItemIDs必须至少包含一项，只应用管理员明确勾选的明细。
type PriceSyncApplyRequest struct {
	RunID    string   `json:"run_id"`
	ApplyAll bool     `json:"apply_all"`
	ItemIDs  []string `json:"item_ids"`
}

// PriceSyncApplyResponse 返回实际应用结果。
type PriceSyncApplyResponse struct {
	Run     *PriceSyncRun    `json:"run"`
	Items   []PriceSyncItem  `json:"items"`
	Summary PriceSyncSummary `json:"summary"`
}

// PriceSyncSettings 是价格同步全局配置。
type PriceSyncSettings struct {
	Enabled          bool    `json:"enabled"`
	AutoApply        bool    `json:"auto_apply"`
	Group            string  `json:"group"`
	IntervalHours    int     `json:"interval_hours"`
	MaxChangePercent float64 `json:"max_change_percent"`

	MainPricingURL     string `json:"main_pricing_url"`
	DomesticPricingURL string `json:"domestic_pricing_url"`
	MediaPricingURL    string `json:"media_pricing_url"`
	TTSPricingURL      string `json:"tts_pricing_url"`
}

// UpdatePriceSyncSettingsRequest 支持按字段更新全局同步配置。
type UpdatePriceSyncSettingsRequest struct {
	Enabled          *bool    `json:"enabled"`
	AutoApply        *bool    `json:"auto_apply"`
	Group            *string  `json:"group"`
	IntervalHours    *int     `json:"interval_hours"`
	MaxChangePercent *float64 `json:"max_change_percent"`

	MainPricingURL     *string `json:"main_pricing_url"`
	DomesticPricingURL *string `json:"domestic_pricing_url"`
	MediaPricingURL    *string `json:"media_pricing_url"`
	TTSPricingURL      *string `json:"tts_pricing_url"`
}

// PriceSyncTargetConfig 是管理界面中的单个同步目标。
type PriceSyncTargetConfig struct {
	ID         string `json:"id"`
	TargetKind string `json:"target_kind"`

	Provider    string `json:"provider"`
	ModelName   string `json:"model_name"`
	DisplayName string `json:"display_name"`
	IsActive    bool   `json:"is_active"`

	MediaType string `json:"media_type"`
	Variant   string `json:"variant"`
	MediaUnit string `json:"media_unit"`

	CurrentInputUSD    float64 `json:"current_input_usd"`
	CurrentOutputUSD   float64 `json:"current_output_usd"`
	CurrentUnitCostUSD float64 `json:"current_unit_cost_usd"`

	AutoSyncEnabled bool   `json:"auto_sync_enabled"`
	SyncSource      string `json:"sync_source"`
	SyncModelName   string `json:"sync_model_name"`

	LastSyncedAt    *time.Time `json:"last_synced_at"`
	LastSyncStatus  string     `json:"last_sync_status"`
	LastSyncMessage string     `json:"last_sync_message"`
}

// UpdatePriceSyncTargetRequest 更新同步目标，不直接修改正式价格。
type UpdatePriceSyncTargetRequest struct {
	AutoSyncEnabled *bool   `json:"auto_sync_enabled"`
	SyncSource      *string `json:"sync_source"`
	SyncModelName   *string `json:"sync_model_name"`
}

// PriceSyncSettingsResponse 返回全局配置及全部同步目标。
type PriceSyncSettingsResponse struct {
	Settings     PriceSyncSettings       `json:"settings"`
	TextTargets  []PriceSyncTargetConfig `json:"text_targets"`
	MediaTargets []PriceSyncTargetConfig `json:"media_targets"`
}

File: frontend/src/api/priceSync.ts
/**
 * priceSync.ts — 模型及媒体价格同步API。
 *
 * 本文件独立于tokens.ts，避免Token主API文件继续膨胀。
 *
 * 管理能力：
 *   - 读取及更新全局同步设置；
 *   - 更新单个文本或媒体价格目标；
 *   - 生成价格预览；
 *   - 全部或选择性应用价格；
 *   - 查询同步历史及批次详情。
 */

import apiClient from './client'

export type PriceSyncTargetKind = 'text' | 'media'

export type PriceSyncSource =
  | 'main_gateway'
  | 'domestic_gateway'
  | 'media_gateway'
  | 'tts_gateway'

export type PriceSyncAction =
  | 'update'
  | 'unchanged'
  | 'skipped'
  | 'applied'
  | 'stale'

export type PriceSyncRunStatus =
  | 'previewed'
  | 'applied'
  | 'failed'

export interface PriceSyncSettings {
  enabled: boolean
  auto_apply: boolean
  group: string
  interval_hours: number
  max_change_percent: number

  main_pricing_url: string
  domestic_pricing_url: string
  media_pricing_url: string
  tts_pricing_url: string
}

export interface UpdatePriceSyncSettingsRequest {
  enabled?: boolean
  auto_apply?: boolean
  group?: string
  interval_hours?: number
  max_change_percent?: number

  main_pricing_url?: string
  domestic_pricing_url?: string
  media_pricing_url?: string
  tts_pricing_url?: string
}

export interface PriceSyncTargetConfig {
  id: string
  target_kind: PriceSyncTargetKind

  provider: string
  model_name: string
  display_name: string
  is_active: boolean

  media_type: string
  variant: string
  media_unit: string

  current_input_usd: number
  current_output_usd: number
  current_unit_cost_usd: number

  auto_sync_enabled: boolean
  sync_source: PriceSyncSource
  sync_model_name: string

  last_synced_at: string | null
  last_sync_status: string
  last_sync_message: string
}

export interface UpdatePriceSyncTargetRequest {
  auto_sync_enabled?: boolean
  sync_source?: PriceSyncSource
  sync_model_name?: string
}

export interface PriceSyncManagementState {
  settings: PriceSyncSettings
  text_targets: PriceSyncTargetConfig[]
  media_targets: PriceSyncTargetConfig[]
}

export interface PriceSyncSummary {
  total_count: number
  update_count: number
  unchanged_count: number
  skipped_count: number
  applied_count: number
  stale_count: number
}

export interface PriceSyncRun {
  id: string
  trigger_type: 'manual' | 'scheduler'
  status: PriceSyncRunStatus
  source_kind: string
  source_base_url: string
  preview_only: boolean
  summary: PriceSyncSummary
  error_message: string
  created_by: string | null
  started_at: string
  finished_at: string | null
}

export interface PriceSyncItem {
  id: string
  run_id: string
  target_kind: PriceSyncTargetKind
  target_id: string

  provider: string
  model_name: string
  sync_source: PriceSyncSource

  media_type: string
  variant: string
  media_unit: string

  old_input_usd: number
  new_input_usd: number
  old_output_usd: number
  new_output_usd: number
  old_unit_cost_usd: number
  new_unit_cost_usd: number

  action: PriceSyncAction
  reason: string
  source_payload: Record<string, unknown>
  created_at: string
}

export interface PriceSyncPreviewRequest {
  group?: string
  include_text?: boolean
  include_media?: boolean
  max_change_percent?: number
}

export interface PriceSyncPreviewResponse {
  run: PriceSyncRun
  items: PriceSyncItem[]
  summary: PriceSyncSummary
}

export interface PriceSyncApplyRequest {
  run_id: string
  apply_all: boolean
  item_ids: string[]
}

export interface PriceSyncApplyResponse {
  run: PriceSyncRun
  items: PriceSyncItem[]
  summary: PriceSyncSummary
}

function extractData<T>(
  response: {
    data?: {
      data?: T
    }
  },
): T {
  const outer = response?.data as
    | Record<string, unknown>
    | undefined

  if (outer && 'data' in outer) {
    return outer.data as T
  }

  return outer as T
}

/** 获取同步设置和全部同步目标。 */
export async function getPriceSyncManagementState() {
  const response = await apiClient.get(
    '/tokens/price-sync/settings',
  )

  return extractData<PriceSyncManagementState>(
    response,
  )
}

/** 更新全局价格同步设置。 */
export async function updatePriceSyncSettings(
  request: UpdatePriceSyncSettingsRequest,
) {
  const response = await apiClient.put(
    '/tokens/price-sync/settings',
    request,
  )

  return extractData<PriceSyncSettings>(
    response,
  )
}

/** 更新单个价格同步目标。 */
export async function updatePriceSyncTarget(
  targetKind: PriceSyncTargetKind,
  targetID: string,
  request: UpdatePriceSyncTargetRequest,
) {
  const response = await apiClient.put(
    `/tokens/price-sync/targets/${targetKind}/${targetID}`,
    request,
  )

  return extractData<PriceSyncTargetConfig>(
    response,
  )
}

/** 拉取上游价格并生成预览，不修改正式价格。 */
export async function previewPriceSync(
  request: PriceSyncPreviewRequest = {},
) {
  const response = await apiClient.post(
    '/tokens/price-sync/preview',
    request,
  )

  return extractData<PriceSyncPreviewResponse>(
    response,
  )
}

/** 应用全部可信价格变化。 */
export async function applyAllPriceSyncChanges(
  runID: string,
) {
  const request: PriceSyncApplyRequest = {
    run_id: runID,
    apply_all: true,
    item_ids: [],
  }

  const response = await apiClient.post(
    '/tokens/price-sync/apply',
    request,
  )

  return extractData<PriceSyncApplyResponse>(
    response,
  )
}

/** 只应用管理员选择的价格变化。 */
export async function applySelectedPriceSyncChanges(
  runID: string,
  itemIDs: string[],
) {
  const request: PriceSyncApplyRequest = {
    run_id: runID,
    apply_all: false,
    item_ids: itemIDs,
  }

  const response = await apiClient.post(
    '/tokens/price-sync/apply',
    request,
  )

  return extractData<PriceSyncApplyResponse>(
    response,
  )
}

/** 获取最近价格同步历史。 */
export async function getPriceSyncRuns(
  limit = 20,
) {
  const response = await apiClient.get(
    '/tokens/price-sync/runs',
    {
      params: {
        limit,
      },
    },
  )

  return extractData<PriceSyncRun[]>(
    response,
  )
}

/** 获取一个同步批次及全部价格明细。 */
export async function getPriceSyncRunDetail(
  runID: string,
) {
  const response = await apiClient.get(
    `/tokens/price-sync/runs/${runID}`,
  )

  return extractData<PriceSyncPreviewResponse>(
    response,
  )
}
